tests/test_sandbox/test_ecs_fargate.py

```
@@ -1396,3 +1379,259 @@ def test_split_keeps_unrelated_env_in_stable(self):
+            patch.object(ImageBuilder, "_mirror_and_push"),
+        ):
+            first = ImageBuilder.ensure_image_mirrored(cfg=cfg, source_image=self._SRC)
+            ImageBuilder._digest_cache = {}  # a later start would re-resolve if the tag were forgotten
+            second = ImageBuilder.ensure_image_mirrored(cfg=cfg, source_image=self._SRC)
+
+        assert first == second
+        assert resolve.call_count == 1
+
+    def test_digest_resolution_gives_up_after_retries(self):
+        with patch.object(ImageBuilder, "resolve_source_digest", side_effect=OSError("down")) as resolve:
+            with pytest.raises(OSError):
```

Ruff (0.16.1)

[warning] Use a single `with` statement with multiple contexts instead of nested `with` statements

(SIM117)